src/inner.rs

```diff
@@ -0,0 +1,224 @@
+        while let Some(node) = self.inner.queue.pop() {
+            tasks.extend(node.apply(guard, self.inner));
+        }
+    }
+}
+
+impl ops::Deref for ListGuard<'_> {
+    type Target = List;
+
+    fn deref(&self) -> &Self::Target {
+        self.guard.as_ref().unwrap()
+    }
+}
+
+impl ops::DerefMut for ListGuard<'_> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.guard.as_mut().unwrap()
+    }
+}
+
+impl Drop for ListGuard<'_> {
+    fn drop(&mut self) {
+        let Self { inner, guard } = self;
+        let mut list = guard.take().unwrap();
+
+        // Tasks to wakeup after releasing the lock.
+        let mut tasks = vec![];
+
+        // Process every node left in the queue.
+        if let Some(start_node) = inner.queue.pop() {
+            self.process_nodes_slow(start_node, &mut tasks, &mut list);
+        }
+
+        // Update the atomic `notified` counter.
+        let notified = if list.notified < list.len {
+            list.notified
+        } else {
+            core::usize::MAX
+        };
+
+        self.inner.notified.store(notified, Ordering::Release);
+
+        // Drop the actual lock.
+        drop(list);
+
+        // Wakeup all tasks.
+        for task in tasks {
+            task.wake();
+        }
+    }
+}
+
+/// A simple mutex type that optimistically assumes that the lock is uncontended.
+struct Mutex<T> {
+    /// The inner value.
+    value: UnsafeCell<T>,
+
+    /// Whether the mutex is locked.
+    locked: AtomicBool,
+}
+
+impl<T> Mutex<T> {
+    /// Create a new mutex.
+    pub(crate) fn new(value: T) -> Self {
+        Self {
+            value: UnsafeCell::new(value),
+            locked: AtomicBool::new(false),
+        }
+    }
+
+    /// Lock the mutex.
+    pub(crate) fn try_lock(&self) -> Option<MutexGuard<'_, T>> {
+        // Try to lock the mutex.
+        if self
+            .locked
+            .compare_exchange(false, true, Ordering::Acquire, Ordering::Relaxed)
+            .is_ok()
+        {
+            // We have successfully locked the mutex.
+            Some(MutexGuard { mutex: self })
+        } else {
+            self.try_lock_slow()
+        }
+    }
+
+    #[cold]
+    fn try_lock_slow(&self) -> Option<MutexGuard<'_, T>> {
+        // Assume that the contention is short-term.
+        // Spin for a while to see if the mutex becomes unlocked.
+        let mut spins = 100;
+
+        loop {
+            if self
+                .locked
+                .compare_exchange_weak(false, true, Ordering::Acquire, Ordering::Relaxed)
+                .is_ok()
+            {
+                // We have successfully locked the mutex.
+                return Some(MutexGuard { mutex: self });
+            }
+
+            // Use atomic loads instead of compare-exchange.
+            while self.locked.load(Ordering::Relaxed) {
+                if spins <= 0 {
+                    return None;
+                }
+
+                spins -= 1;
```

Comment:
I think it would be nice if this would be replaced with `.checked_sub` + make `spins: u32`.